lib/Codegen/Lowering/Distance/EditDistance.cpp

```diff
@@ -0,0 +1,73 @@
+    const unsigned int n = expected.length();
+
+    // Create a matrix to store partial distances for the dynamic programming algorithm.
+    std::vector<unsigned int> distances((n+1)*(m+1), 0);
+
+    // Initialize the first row of the matrix (all deletions).
+    for (unsigned int j=0; j<=n; ++j) {
+      distances[j] = j*deletion_cost;
+    }
+
+    // Initialize the first column of the matrix (all insertions).
+    const unsigned int first_char_idx = (n > 0) ? charToIndex(expected[0]) : 0;
+    for (unsigned int i=1; i<=m; ++i) {
+      const unsigned int actual_char_idx = charToIndex(actual[i-1]);
+      const unsigned int char_insertion_cost = (n > 0) ?
+                                          substitution_costs[actual_char_idx][first_char_idx] :
+                                          largest_substitution_cost;
+      distances[i*(n+1)] = distances[(i-1)*(n+1)] + base_insertion_cost + char_insertion_cost;
+    } 
+
+    // Compute the distance matrix.
+    for (unsigned int i=1; i<=m; ++i) {
+      for (unsigned int j=1; j<=n; ++j) {
+        const unsigned int index = i*(n+1) + j;
+        const unsigned int actual_char_idx = charToIndex(actual[i-1]);
+        const unsigned int expected_char_idx = charToIndex(expected[j-1]);
+
+        // Calculate the distance in case of deletion.
+        const unsigned int deletion_result = distances[index - 1] + deletion_cost;
+
+        // Calculate the distance in case of insertion.
+        const unsigned int current_char_cost = substitution_costs[actual_char_idx][expected_char_idx];
+        const unsigned int next_char_cost = (j < n) ?
+              substitution_costs[actual_char_idx][charToIndex(expected[j])] :
+              largest_substitution_cost;
+        const unsigned int insertion_result = distances[index - (n+1)] + base_insertion_cost + 
+                                              std::min(current_char_cost, next_char_cost);
+
+        // Calculate the distance in case of substitution.
+        const unsigned int substitution_result = distances[index - (n+1) - 1] + 
+                                                 substitution_costs[actual_char_idx][expected_char_idx];
+
+        // Calculate the distance in case of transposition.
+        const unsigned int transposition_result = 
+                           (i >= 2 && j >= 2 && actual[i-1] == expected[j-2] && actual[i-2] == expected[j-1]) ?
+                           distances[index - 2*(n+1) - 2] + transposition_cost :
+                           max_cost;
+
+        // Select the operation with the lowest cost.
+        distances[index] = std::min(std::min(deletion_result, insertion_result), 
+                                    std::min(substitution_result, transposition_result));
+      }
+    }
+
+    return distances[(n+1)*(m+1)-1];
+  }
+
+  const unsigned int EditDistance::largest_substitution_cost = 21;
+  // Values obtained with a python script, using a measure of keyboard distance and
```

Comment:
This needs to be explained way better. Which Python script? What does it do? Why 2 as penalty?

Reply:
The python script I am referring to is not in the repository. It is a simple script in which I wrote a simplified representation of a keyboard with the classic QWERTY layout, using a misclassification cost proportional to the physical distance between keys. The value of 2 has no mathematical meaning, it was simply found to work well experimentally. For a heuristic reason, 2 is also the cost for a mismatch for keys that adjacent in the same row. I will improve the comment.